docs/designs/anonlink-upload-data-plan.md

```diff
@@ -0,0 +1,289 @@
+pulled by the server. Modifying the client and API to upload in chunks - making many smaller upload requests - was also
+considered. This alternative was rejected as we estimate it would take more effort to get to the same reliability. For
+example the chunking, compression, retrying, and data integrity would all have to be custom handled.
+
+With the Anonlink system we aim to provide everything required to take users from clean records through to an accurate
+concordance table. We rejected the idea of requiring a data provider to setup their own object store or otherwise
+hosting their encoded data online; instead we will modify the client to be able to upload to an object store.
+
+Not including the initial design, implementation is expected to be approximately 2 sprints of work for 2 software 
+engineers. The high level objective is that Anonlink be usable to link between datasets that are roughly the size of 
+the Australian population. 
+
+The aim of the rest of this document is to specify the design of the "pull data" feature.
+
+The sections are:
+
+- Requirements
+- High level system design
+- Specifications for implementation (including the REST API).
+- decisions and alternatives considered. What would be the "best"/"easiest" approach.
+
+
+## Requirements
+
+- A method for requesting temporary & restricted object store credentials shall be added to the Anonlink Entity Service 
+  API. The temporary credentials will authorize the holder to upload data, the credentials **must not** provide access 
+  to any other data providers encodings, or other working data stored in the object store.
+- A method for a data provider to provide an object store URI pointing to the encoding data instead of uploading it 
+  shall be added to the Anonlink Entity Service API. This method **may** support pulling directly via http as well.
+- A deployment option to expose the MinIO object store via an ingress will be added. As this increases the attack 
+  surface an internal security review of MinIO must be conducted. Deployment using default or publicly committed 
+  credentials must be mitigated.
+- The maximum size of a clients' encodings will be 1TiB.
+
+
+## High Level Design
+
+The Anonlink Entity Service already relies on MinIO (or AWS S3) as an ancillary services for storing encoding data
+as well as data for in progress computations. MinIO can easily and robustly handle the expected volume of data, and we 
+will leverage the object store to support this pull feature.
+
+At deployment time the user will have the option of exposing the MinIO object store used by the Anonlink Entity Service,
+or - as already supported - providing external S3/MinIO access credentials. The Anonlink Entity Service's API will be
+modified such that a data provider can request temporary object store credentials. The client uses these temporary
+credentials to upload their encodings to the object store. The client then informs the Anonlink server via the `clks` endpoint
```

Comment:
That could work too. 

It could involve some trickery to create Minio users dynamically, the [docs](https://github.com/minio/minio/tree/master/docs/multi-user) only show creating users/groups via the `mc` command line tool and don't document the underlying API. I was thinking of creating a default upload user using `mc` at service startup, and then using the assume role api to create credentials for each data uploader who requests them.

The [AssumeRole](https://github.com/minio/minio/blob/master/docs/sts/assume-role.md) API only creates temporary credentials (valid between 15min to 12 hours).

From a user point of view I don't see it making a big difference, the majority of users will be using our client tool which can happily make the three calls in sequence. Also this is only for the case where we are exposing our object store, if the user already has object store credentials e.g. for S3 they can upload at their leisure without having to think about temporary credentials. It would be nice for our client tool to deal with this nicely/transparently.

The default case using temporary credentials for our object store:
``` 
$ anonlink upload mydata.csv <AUTH params etc>
```

Where the user wants our client to upload to their own bucket, optionally providing the AWS credential profile to use:
```
$ anonlink upload mydata.csv [--profile easd] --upload-to=s3://my-own-bucket/mydata.csv <AUTH params etc>
```

If the user wants to use already uploaded data, they will have to provide the anonlink entity service with appropriate credentials. I definitely don't want to assume we can share a users AWS credentials with the service, and I'm not sure if we should be asking the users' AWS account for (temporary) credentials, or just require it to be all explicitly provided.

This means something like:
```
$ anonlink upload s3://my-own-bucket/mydata.csv
```
is probably not going to work unless the user has provided us explicit credentials to share with the service, or explicitly told us that our client should create temporary and restricted credentials to share with the service. Telling us could be additional command line arguments, an `~/.anonlink` config file or environment variables. 

cc @wilko77 do you have opinions here?